jetty-cdi/src/main/java/org/eclipse/jetty/cdi/CdiDecorator.java

```diff
@@ -0,0 +1,172 @@
+    private MethodHandle _release;
+
+
+    public CdiDecorator(WebAppContext context) throws ClassNotFoundException, UnsupportedOperationException
+    {
+        _context = context;
+        ClassLoader classLoader = _context.getClassLoader();
+
+        try
+        {
+            _cdiClass = classLoader.loadClass("javax.enterprise.inject.spi.CDI");
+        }
+        catch (Exception e)
+        {
+            throw new UnsupportedOperationException(e);
+        }
+        _beanManagerClass = classLoader.loadClass("javax.enterprise.inject.spi.BeanManager");
+        _annotatedTypeClass = classLoader.loadClass("javax.enterprise.inject.spi.AnnotatedType");
+        _injectionTargetClass = classLoader.loadClass("javax.enterprise.inject.spi.InjectionTarget");
+        _creationalContextClass = classLoader.loadClass("javax.enterprise.context.spi.CreationalContext");
+        _contextualClass = classLoader.loadClass("javax.enterprise.context.spi.Contextual");
+    }
+
+    /**
+     * Decorate an object.
+     * <p>The signature of this method must match what is introspected for by the
+     * Jetty DecoratingListener class.  It is invoked dynamically.</p>
+     *
+     * @param o The object to be decorated
+     * @param <T> The type of the object to be decorated
+     * @return The decorated object
+     */
+    public <T> T decorate(T o)
+    {
+        try
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("decorate {} in {}", o, _context);
+
+            if (_release == null)
+            {
+                MethodHandles.Lookup lookup = MethodHandles.lookup();
+                MethodHandle current = lookup.findStatic(_cdiClass, "current", MethodType.methodType(_cdiClass));
+                MethodHandle getBeanManager = lookup.findVirtual(_cdiClass, "getBeanManager", MethodType.methodType(_beanManagerClass));
+                Object manager = getBeanManager.invoke(current.invoke());
+
+                _createAnnotatedType = lookup.findVirtual(_beanManagerClass, "createAnnotatedType", MethodType.methodType(_annotatedTypeClass, Class.class))
+                    .bindTo(manager);
+                _createInjectionTarget = lookup.findVirtual(_beanManagerClass, "createInjectionTarget", MethodType.methodType(_injectionTargetClass, _annotatedTypeClass))
+                    .bindTo(manager);
+                _createCreationalContext = lookup.findVirtual(_beanManagerClass, "createCreationalContext", MethodType.methodType(_creationalContextClass, _contextualClass))
+                    .bindTo(manager).bindTo(null);
+                _inject = lookup.findVirtual(_injectionTargetClass, "inject", MethodType.methodType(Void.TYPE, Object.class, _creationalContextClass));
+                _dispose = lookup.findVirtual(_injectionTargetClass, "dispose", MethodType.methodType(Void.TYPE, Object.class));
+                _release = lookup.findVirtual(_creationalContextClass, "release", MethodType.methodType(Void.TYPE));
+            }
```

Comment:
I have to do this lazily because there is no fixed order between SCIs.  So when the SCI for this executes and creates the decorator, the SCI that initialises CDI may not yet have been run.  So we can call CDI.current until after all the SCIs have been executed, which is when we first start decorating.

I'll protect this with a sync block

Reply:
I'm still dubious. Here you are looking up things once, bind them, but are we sure thing do not change between invocations? What if the external application changes the `BeanManager` to be another instance? That is apparently possible via API (e.g. `CDI.setCDIProvider()`).

I would rather collect all the `MethodHandle`s in a static block and only invoke them in `decorate()`, passing objects as parameters rather than use `bindTo()`.

Reply:
It can't be a static block as the classloader is different